### PR TITLE
Add 'Implement all' code action

### LINKED ISSUE #30749: Code action for implementing methods list the methods one by one
Consider the following scenario:
```ballerina
class ListenerImpl {
    *Listener;
}

public type Listener object {

    public function 'start() returns error?;

    public function gracefulStop() returns error?;

    public function immediateStop() returns error?;

    public function detach(service object {} s) returns error?;

    public function attach(service object {} s, string[]? name = ()) returns error?;
};
```
When I check the code actions for the above, it gives a code action each for each of the methods. Can we have another option to implement all the methods in one code action?

Screenshot 2021-05-24 at 13 53 31

### PR DESCRIPTION
**Pull request by TharushiJay**
## Purpose
$subject to implement all methods in one code action.

Fixes #30749

## Samples
Supports the same for,

1) Object constructor expressions
Screenshot from 2021-06-28 23-06-02

2) Object types
Screenshot from 2021-06-28 23-09-36

3) Service declarations
Screenshot from 2021-06-28 23-06-25

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Review by IMS94 — CHANGES_REQUESTED** [review]
We have to introduce an intermediate code action provider common to `ImplementAllCodeAction` and `ImplementMethodCodeAction`. Something like `AbstractImplementMethodCodeAction` which extends `AbstractCodeActionProvider`.

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/ImplementAllCodeAction.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+                .filter(diagnostic -> DIAGNOSTICCODES.contains(diagnostic.diagnosticInfo().code()))
+                .collect(Collectors.toList());
+        
+        if (diags.isEmpty() || diags.size() == 1) {
+            return Collections.emptyList();
+        }
+
+        SyntaxTree syntaxTree = context.workspace().syntaxTree(context.filePath()).orElseThrow();
+        ImportsAcceptor importsAcceptor = new ImportsAcceptor(context);
+        List<TextEdit> edits = new ArrayList<>(importsAcceptor.getNewImportTextEdits());
```
Comment (IMS94) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/ImplementAllCodeAction.java:84]: We don't need the imports acceptor here right?

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/AbstractImplementMethodCodeAction.java [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+import static org.ballerinalang.langserver.common.utils.CommonUtil.LINE_SEPARATOR;
+
+/**
+ * Represents the common class for implementing the functions of an object.
+ *
+ * @since 2.0.0
+ */
+@JavaSPIService("org.ballerinalang.langserver.commons.codeaction.spi.LSCodeActionProvider")
+public abstract class AbstractImplementMethodCodeAction extends AbstractCodeActionProvider {
```
Comment (IMS94) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/AbstractImplementMethodCodeAction.java:60]: We can't have this annotation here

**Review by IMS94 — CHANGES_REQUESTED** [review]
Overall looks good. Need 2 modifications.

**Review by malinthar — DISMISSED** [review]
LGTM

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/AbstractImplementMethodCodeAction.java [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+/**
+ * Represents the common class for implementing the functions of an object.
+ *
+ * @since 2.0.0
+ */
+public abstract class AbstractImplementMethodCodeAction extends AbstractCodeActionProvider {
+    protected static final int DIAG_PROPERTY_NAME_INDEX = 0;
+    protected static final int DIAG_PROPERTY_SYMBOL_INDEX = 1;
+
+    protected static final Set<String> DIAGNOSTICCODES = Set.of(
```
Comment (nadeeshaan) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/AbstractImplementMethodCodeAction.java:62]: Use `DIAGNOSTIC_CODES`

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/AbstractImplementMethodCodeAction.java [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+    }
+
+    /**
+     * Returns a list of text edits based on diagnostics.
+     *      *
+     *      * @param diagnostic diagnostic to evaluate
+     *      * @param positionDetails   {@link DiagBasedPositionDetails}
+     *      * @param context    language server context
+     *      * @return list of Text Edits
+     */
```
Comment (nadeeshaan) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/AbstractImplementMethodCodeAction.java:91]: Fix the formatting here

**Review by nadeeshaan — CHANGES_REQUESTED** [review]

**Comment by nadeeshaan** [comment]
Here, when we are generating a function with a return, shall we try to include the return statement as well? Otherwise there is a semantic error in the generated code segment.

**Review by nadeeshaan — APPROVED** [review]
LGTM

**Review by IMS94 — APPROVED** [review]
LGTM!

**Comment by TharushiJay** [comment]
> Here, when we are generating a function with a return, shall we try to include the return statement as well? Otherwise there is a semantic error in the generated code segment.

Created issue #31608 for this.

**Pull request merged**